### ISSUE TITLE
DNS via WireGuard in advanced mode

### ISSUE THREAD

**Issue by Terrance** [celzero/rethink-app#1510]
In Simple mode for WireGuard, all DNS requests are routed via the proxy.  In Advanced mode, the main DNS choice applies for apps both included and excluded in the routing (according to Logs > DNS and the description of advanced mode), and there's no option to use WireGuard for DNS explicitly.

I'd expect/like DNS requests for included apps to also be routed through the proxy.

Context: I'm using a WireGuard proxy in order to access local web services in a handful of apps, but I need the DNS to resolve for those apps at least (local DNS resolves additional subdomains of a "real" domain to internal IPs).

**Comment by ignoramous** [comment]
This inability to per-app split tunnel DNS is an Android limitation (Android forwards DNS to Rethink on behalf of all apps and hence to Rethink it appears as if all DNS queries were sent by the OS).

We did lay the ground work in the network engine to do split tunneling in userspace (#270) but it is too complicated to handle all use-case (the UI would need just too many knobs).

**Comment by Terrance** [comment]
Would an option to use WireGuard for all DNS whilst in advanced mode be feasible?

**Comment by ignoramous** [comment]
> Would an option to use WireGuard for all DNS whilst in advanced mode be feasible?

As in, redirect ALL DNS queries through to a particular WireGuard's DNS? Yes, it is possible. We'll see if we can do this. Ideally, we'd want to complete our implementation of per-app split-tunnel DNS itself. There's so much to do, and not enough time.

We are also thinking about letting users "pin" domain names to preset DNS upstreams: 
- #1153

**Comment by Terrance** [comment]
> As in, redirect ALL DNS queries through to a particular WireGuard's DNS?

Yes -- I'm running [Blocky DNS](https://0xerr0r.github.io/blocky/) with an internal zone for a service domain, so being able to use that via WG to benefit from the block lists / custom DNS entries would be nice.

> letting users "pin" domain names to preset DNS upstreams

That would also work in this case (assuming WireGuard is one of the choices), so I've added a +1 there.

**Comment by funkypunkyz** [comment]
These DNS leaks currently also happen with Orbot, if I am not mistaken.
For example visiting `https://check.torproject.org` with a browser app proxied through Orbot will leak a DNS request for `check.torproject.org` to the resolver outside Orbot (System DNS,...), and you can see this request in the DNS tab. IP packet request then goes through Orbot.

Wouldn't it be good idea to add the same DNS warning as with Wireguard for now?

> DNS is forwarded to `<user-preferred DNS upstream ouside Orbot>....`

**Comment by ignoramous** [comment]
>  For example visiting https://check.torproject.org with a browser app proxied through Orbot will leak a DNS request for check.torproject.org to the resolver outside Orbot (System DNS,...), and you can see this request in the DNS tab. IP packet request then goes through Orbot.

It cannot be considered a leak unless Rethink missed processing the DNS query and it somehow was instead resolved by resolver of the underlying network (ex: Wifi / Mobile).

- Is Rethink forwarding `.onion` domains to "System DNS"? If so, we've disabled this behaviour: #1259
- Is Rethink forwarding all domains to "System DNS"? If so, is "System DNS" is what you're using? If *not*, it can be considered a leak.

> Wouldn't it be good idea to add the same DNS warning as with Wireguard on the Orbot UI widget for now?

One has to setup Rethink to forward DNS to Orbot. Orbot is a TCP-only proxy (and that's mentioned in the setup UI). Unfortunately, Orbot does not provide a *stable* API to integrate with its DNS endpoint "automatically", and hence that step remains manual. Besides, Orbot integration provides per-app proxying capabilities (which is similar to running WireGuard in "Advanced" mode).

**Comment by funkypunkyz** [comment]
Thanks @ignoramous for taking the time. I don't have deep knowledge of technical implementation in Rethink, will try my best to answer from user's perspective.

> It cannot be considered a leak unless Rethink missed processing the DNS query and it somehow was instead resolved by resolver of the underlying network (ex: Wifi / Mobile).

Yeah, under DNS tab I saw the per-app Orbot proxied request, hence Rethink in principle seems able to process it. Effectively it then routed this request to "System DNS". I also re-checked that it arrived at my network DNS resolver in home network outside Android OS. If you call it "leak" or not, it seems unexpected from user perspective.

> Is Rethink forwarding .onion domains to "System DNS"? If so, we've disabled this behaviour: Routing of .onion DNS Querys diffrent as before? #1259

Didn't try `.onion` domains, as not important for me atm.

> Is Rethink forwarding all queries to "System DNS"? If so, is "System DNS" is what you're using? If not, it can be considered a leak.
 
Yes, "System DNS" is, what I have configured as DNS proxy type in Rethink. My assumption was, that it  would be used only for non-proxied traffic.

> One has to setup Rethink to forward DNS to Orbot. Orbot is a TCP-only proxy (and that's mentioned in the setup UI). Unfortunately, Orbot does not provide a stable API to integrate with its DNS endpoint "automatically", and hence that step remains manual. 

To understand that a bit better for myself (please correct otherwise): 
(1) Under Proxy -> Setup Orbot you can configure per-app proxying settings for Orbot, where traffic is automatically proxied for selected apps, either via HTTP Proxy or TCP-only Proxy.
(2) Under Proxy -> Other, you can configure a manual endpoint, where Orbot listens on incoming requests, either on SOCKS5 or as HTTP CONNECT proxy. Apps need to support and be configured via manual proxy settings. 

(Btw: For my tests I chose (1) with  the `CONNECT` method of HTTP proxy.)

Did you mean (2) by "setup Rethink to forward DNS to Orbot" / "step remains manual"? If not: Are there instructions for this manual setup available?

> Besides, Rethink's Orbot integration provides per-app proxying capabilities (which is similar to running WireGuard in "Advanced" mode) but per-app DNS isn't possible on Android (without painful, possibly tricky changes on our end, which this github issue tracks).

Am I right, there is no "simple mode" for Orbot - you either have per-app proxying (1) or manual proxying (2)? And currently Orbot is leaky per default with (1)? I have not tried option (2) yet, is there anything to consider when using Rethink in VPN Lockdown mode or should this work without leaks?

Overall, your your comment unfortunately sounds, as if strict tunneling is hard to achieve with Android's architecture and due to the way DNS requests flow. If I am not mistaken, GrapheneOS developers are currently working on better VPN leak prevention. Might it help here to communicate for better integration between Android (AOSP) <-> Rethink?

Sorry for these many questions. It is is a very interesting topic, though takes a bit time for me to grasp all available options in Rethink. Thanks you for such a great product!

**Comment by ignoramous** [comment]
Thanks.

>  If you call it "leak" or not, it seems unexpected from user perspective.

The problem with inaccurately using *technical* terminology is it ends up being a cause of confusion. What you're observing isn't a *DNS leak* even if it "seems" like it is one. 

> Yes, "System DNS" is, what I have configured as DNS proxy type in Rethink. My assumption was, that it would be used only for non-proxied traffic.

Two things here:
- For SOCKS and HTTP CONNECT tunnels, user-preferred DNS upstream (as set in *Configure* -> *DNS* is used). This upstream is *proxied* over / *tunneled* within the active SOCKS/HTTP CONNECT endpoint unless,
  1. *Configure* -> *DNS* -> *Do not proxy DNS* is turned ON.
  2. Or: the upstream is NOT *System DNS*.
- For WireGuard in *Advanced* mode and *Orbot* (one-click Orbot integration in Rethink), unless we impl fixes for this current github issue, Rethink *will not* do per-app DNS, and instead continue to use user-preferred DNS upstream (as set in *Configure* -> *DNS*).
- For WireGuard in *Simple* mode, DNS queries from ALL apps are sent to DNS upstream specified in the active WireGuard configuration.

> Did you mean (2) by "setup Rethink to forward DNS to Orbot" / "step remains manual"? If not: Are there instructions for this manual setup available?

You can try choosing "Orbot" in *Configure* -> *DNS* -> *Other DNS* -> *DNS 53*. Doing so will send DNS queries from ALL apps to Orbot (if it works).

> If I am not mistaken, GrapheneOS developers are currently working on better VPN leak prevention. Might it help here to communicate for better integration between Android (AOSP) <-> Rethink?

That's a different issue that they're trying to solve (a fix that Rethink doesn't even need).

**Comment by funkypunkyz** [comment]
> The problem with inaccurately using technical terminology is it ends up being a cause of confusion. What you're observing isn't a DNS leak even if it "seems" like it is one.

Not sure what the appropriate technical definition for DNS leak is. From user's perspective I would say: DNS traffic forwarded in unexpected ways, resulting in a privacy leak.  See also:

[Wikipedia](https://en.wikipedia.org/wiki/DNS_leak)
> A DNS leak is a security flaw that allows DNS requests to be revealed to ISP DNS servers, despite the use of a VPN service to attempt to conceal them.

www.dnsleaktest.com
> What is a DNS leak and why should I care?
Under certain conditions, even when connected to the anonymity network, the operating system will continue to use its default DNS servers instead of the anonymous DNS servers assigned to your computer by the anonymity network. DNS leaks are a major privacy threat since the anonymity network may be providing a false sense of security while private data is leaking.

Though Wikipedia article states, DNS leaks only happen with Split-Tunnel VPN (not Lockdown mode), which is an argument in your favor. Note I do not want to be nitpicky here, just trying to understand your technical point of view.

> Two things here:

First point made me look at Configure -> Proxy -> Other setting again. My understanding now is, that it is a mutual exclusive upstream alternative in addition to Orbot or Wirguard, where traffic of one specific app is forwarded to an upstream custom SOCKS/HTTP proxy, right? I probably mixed things up a bit in previous post.

> You can try choosing "Orbot" in Configure -> DNS -> Other DNS -> DNS 53. Doing so will send DNS queries from ALL apps to Orbot (if it works).

Thanks! Will try it. This was a bit harder to find.

As general suggestion for UI structure: Wouldn't it be useful to make "System DNS", "Other DNS", "Rethink DNS" settings (Configure -> DNS) only usable in combination with unproxied traffic? My own assumption is, that from privacy perspective it doesn't make sense to route traffic through a different DNS server than the one provided by enabled VPN/Tor/anonymity network.

> That's a different issue that they're trying to solve (a fix that Rethink doesn't even need).

Good to know.

**Comment by ignoramous** [comment]
"despite the use of a VPN service to attempt to conceal them."

Rethink isn't attempting to conceal the requests it by itself sent to *System DNS*.

For Rethink to send DNS queries to Orbot (per-app), 2 things need to happen:
1. This current issue must be fixed (to overcome Android's limitation)
2. Orbot needs to start and expose its DNS endpoint to apps like Rethink.

Point 2 above is something I never see happening (#665), so either this "leak" will have to be plugged by setting up DNS 53 to connect to Orbot or not use Orbot with Rethink.

**Comment by funkypunkyz** [comment]
Hm OK I see. The most practical solution I can think of is to have separate user profiles:
- one for apps that depend on Rethink's DNS/firewall/Wireguard tooling - last hopefully with proper DNS tunneling over VPN
- the other for Orbot in VPN lockdown mode, either routing specific apps or all over Tor

Switching between Rethink standalone and Rethink+upstream Orbot by visiting two different app settings would be too unpractical and error-prone. Maybe add an additional setting or link in the Orbot one-click widget to switch to DNS 53 for leak prevention?

**Comment by ignoramous** [comment]
> one for apps that depend on Rethink's DNS/firewall/Wireguard tooling - last hopefully with proper DNS tunneling over VPN
> And if I understood that, there is still a chance to fix Wireguard DNS fully, which might be the more frequent use case.

Yep.

To be clear, in *Simple* mode, Rethink already has "proper DNS tunneling". *Advanced* mode is for power users to run multiple WireGuards at once (which no other WireGuard app on Android supports. Even if split-tunneling DNS is "solvable" to some extent, I consider it a niche use-case, not a "frequent" one).

>  Maybe add an additional setting or link in the Orbot one-click widget to switch to DNS 53 for leak prevention or prompt user for automatic adjustment?

We had this. It didn't work always for all users causing more frustration (#665). So, now we don't. Orbot needs to fix this, but I am unsure if they ever will.

You may also try using anonymizing DNS protocols like *ODoH* or *DNSCrypt v3* (with *Relays*) when using Orbot, as a different workaround.

> the other for Orbot in VPN lockdown mode, either routing specific apps or all over Tor

Yeah, Orbot running as a "proxy" is not the same as it running as a "VPN", which may lead to surprises (ex: "DNS leaks") for those who are unaware.

**Comment by funkypunkyz** [comment]
> To be clear, in Simple mode, Rethink already has "proper DNS tunneling".

Yes, I know, But this will tunnel every traffic from OS, which might be not desirable either.

> Advanced mode is for power users to run multiple WireGuards at once

I had a different use case in mind: running only specific apps through Wireguard tunnel, while routing all others through System DNS - all in a single user profile. For example using Youtube app or Play Store via VPN IP with privacy in mind.

> We had this. It didn't work always for all users causing more frustration (https://github.com/celzero/rethink-app/issues/665). So, now we don't. Orbot needs to fix this, but I am unsure if they ever will.

Makes sense. Unfortunately I cannot comment on technical details here.

> Even if split-tunneling DNS is "solvable" to some extent, I consider it a niche use-case, not a "frequent" one).

I am not sure of this one.
- Need for a VPN often is app-specific (e.g. above examples).
- I think, Rethink users on average are already more technically affine and "advanced".
- Empirically speaking (though only a couple of people) - need for VPN per app seemed to be one of the more common requirements.
- There will come up need to compartmentalize network traffic flexibly with privacy in mind.

> Yeah, Orbot running as a "proxy" is not the same as it running as a "VPN", which may lead to surprises (ex: "DNS leaks") for those who are unaware.

So running Orbot as SOCKS/HTTP CONNECT proxy is leak-safe, while in VPN mode might not be fully leak-proof? IIRC, this VPN leak is an Android OS issue. Perhaps GrapheneOS devs might be able to improve this.

**Comment by ignoramous** [comment]
You make good points. 

(we have done a LOT of work already to make per-app DNS work, let's see if we mange to ship it *fully* within the next few months).

**Comment by funkypunkyz** [comment]
Awesome, thank you very much for your work!

**Comment by funkypunkyz** [comment]
@ignoramous  Does anything speak against using RethinkDNS in combination with Android [Private spaces](https://source.android.com/docs/security/features/private-space) (Android 15+) from your side?

Private spaces allow for independent VPN connections. So in order to prevent DNS leaks, we might just install an app requiring VPN together with a separate RethinkDNS instance + "Simple mode"  in a private space. This sounds like a system-enforced, great alternative to workarounds needed due to Android VPN architecture specifics.

One disadvantage certainly is, that you need to re-configure Rethink for each space, but this probably is managable via export/import settings and space-specific post-customization.

I am not sure, how "leak-proof" you would assess current implementation covered by this issue. If this turns out to be not reliable enough, an idea would be to disable "Advanced mode"/app-dependent VPN mode for Android 15+ and instead allow a simple way to generate a private space for given app + Rethink.

**Comment by funkypunkyz** [comment]
Oh, I wrongly assumed there would be multiple private space**s** possible - sorry for the confusion. With only one private space the idea is not scalable. Hence OP feature request is still relevant.

**Comment by ignoramous** [comment]
Per-app DNS, which forms the basis for DNS resolution via WireGuard, has shipped in `v055o`: https://github.com/celzero/rethink-app/releases/tag/v0.5.5o

Try it out and let us know how it goes! Closing this bug, but feel free to ask to reopen it.

---

From https://github.com/celzero/rethink-app/issues/1872#issuecomment-3169574690:

On Android 12+, Rethink `v055o` onwards will now show the app name in *Configure* -> *Logs* -> *DNS* (tab).

On Android 11 & below, end users will have to turn ON from *Configure* -> *DNS*:
1. *Advance DNS filtering* &
2. *Split DNS*.


https://github.com/celzero/firestack/issues/136
